packages/database/src/changeHandlers/TaskCreationHandler.js

```diff
@@ -0,0 +1,107 @@
+/*
+ * Tupaia
+ *  Copyright (c) 2017 - 2024 Beyond Essential Systems Pty Ltd
+ */
+import keyBy from 'lodash.keyby';
+import { ChangeHandler } from './ChangeHandler';
+
+const getAnswerWrapper = (config, answers) => {
+  const answersByQuestionId = keyBy(answers, 'question_id');
+
+  return questionKey => {
+    const questionId = config[questionKey]?.questionId;
+    if (!questionId) {
+      return null;
+    }
+    const answer = answersByQuestionId[questionId];
+    return answer?.text;
+  };
+};
+
+const isPrimaryEntityQuestion = (config, questions) => {
+  const primaryEntityQuestion = questions.find(question => question.type === 'PrimaryEntity');
+  const { questionId } = config['entityId'];
+  return primaryEntityQuestion.id === questionId;
+};
+
+const getSurveyCode = async (models, config) => {
+  const surveyCode = config.surveyCode;
+  const survey = await models.survey.findOne({ code: surveyCode });
+  return survey.id;
+};
+
+const getQuestions = (models, surveyId) => {
+  return models.database.executeSql(
+    `
+        SELECT q.*, ssc.config::json as config
+        FROM question q
+        JOIN survey_screen_component ssc ON ssc.question_id = q.id
+        JOIN survey_screen ss ON ss.id = ssc.screen_id
+        WHERE ss.survey_id = ?;
+      `,
+    [surveyId],
+  );
+};
+
+export class TaskCreationHandler extends ChangeHandler {
+  constructor(models) {
+    super(models, 'task-creation-handler');
+
+    this.changeTranslators = {
+      surveyResponse: change => this.getNewSurveyResponses(change),
+    };
+  }
+
+  /**
+   * @private
+   * Only get the new survey responses that are created, as we only want to mark tasks as completed when a survey response is created, not when it is updated
+   */
+  getNewSurveyResponses(changeDetails) {
+    const { type, new_record: newRecord, old_record: oldRecord } = changeDetails;
+
+    // if the change is not a create, we don't need to do anything. This is because once a task is marked as complete, it will never be undone
+    if (type !== 'update' || !!oldRecord) {
+      return [];
+    }
+    return [newRecord];
+  }
+
+  async handleChanges(models, changedResponses) {
+    // if there are no changed responses, we don't need to do anything
+    if (changedResponses.length === 0) return;
+
+    for (const response of changedResponses) {
+      const sr = await models.surveyResponse.findById(response.id);
+      const questions = await getQuestions(models, response.survey_id);
+
+      const taskQuestion = questions.find(question => question.type === 'Task');
```

Comment:
What would happen if there were multiple task questions? Technically this is allowed, as there is not validation to stop this happening

Reply:
thanks for raising that - I hadn't considered that. I have updated the change handler to handle multiple task questions